### PR TITLE
8252689: Classes are loaded from jrt:/java.base even when CDS is used

### PR DESCRIPTION
**Pull request by yminqi**
Java.util.jar.Manifest related classes not archived since they are neither in classlist or loaded in dump process. Manually create a dummy Manifest object will cause those classes loaded in dump and archived.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252689](https://bugs.openjdk.java.net/browse/JDK-8252689): Classes are loaded from jrt:/java.base even when CDS is used

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to ee270886df940a093e5f551d4697a1e9ac894983
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to ee270886df940a093e5f551d4697a1e9ac894983

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/134/head:pull/134`
`$ git checkout pull/134`

### CODE REVIEW

**Comment by cl4es** [comment]
I think you could achieve the same if you extend the [HelloClasslist](https://github.com/openjdk/jdk/blob/master/make/jdk/src/classes/build/tools/classlist/HelloClasslist.java) tool built and used to generate the default classlist file to include a manifest in its jar file. Add a dummy manifest.mf somewhere and edit the CLASSLIST_JAR target in [GenerateLinkOptData](https://github.com/openjdk/jdk/blob/master/make/GenerateLinkOptData.gmk) to include it in the jar.

This would avoid hard-coding  this in hotspot code and keep the responsibility of curating the default classlist in one place.

**Comment by iklam** [comment]
> I think you could achieve the same if you extend the [HelloClasslist](https://github.com/openjdk/jdk/blob/master/make/jdk/src/classes/build/tools/classlist/HelloClasslist.java) tool built and used to generate the default classlist file to include a manifest in its jar file. Add a dummy manifest.mf somewhere and edit the CLASSLIST_JAR target in [GenerateLinkOptData](https://github.com/openjdk/jdk/blob/master/make/GenerateLinkOptData.gmk) to include it in the jar.
> 
> This would avoid hard-coding this in hotspot code and keep the responsibility of curating the default classlist in one place.

During the JDK build, HelloClasslist is already packaged with a manifest:

```
$ jar tf ./support/classlist.jar
META-INF/
META-INF/MANIFEST.MF
build/tools/classlist/HelloClasslist.class
```
And the default CDS archive contains the 3 classes mentioned in the bug report:
```
$ ./images/jdk/bin/java -cp ./support/classlist.jar -verbose build.tools.classlist.HelloClasslist | \
    egrep '((java.util.jar.Attributes)|(java.util.LinkedHashMap)|(java.util.jar.Manifest.FastInputStream)) source'
[0.044s][info][class,load] java.util.jar.Attributes source: shared objects file
[0.044s][info][class,load] java.util.LinkedHashMap source: shared objects file
[0.044s][info][class,load] java.util.jar.Manifest$FastInputStream source: shared objects file
```
However, for some reason, if you run a simple HelloWorld.jar to collect the class list, these 3 classes are not loaded. As a result, these classes may not be stored in custom archives created from your own class lists.
```
$ jar tf ~/tmp/HelloWorld.jar
META-INF/
META-INF/MANIFEST.MF
HelloWorld.class
$ ./images/jdk/bin/java -Xshare:off -cp ~/tmp/HelloWorld.jar -verbose \
    -XX:DumpLoadedClassList=foo.list HelloWorld | egrep \
    '((java.util.jar.Attributes)|(java.util.LinkedHashMap)|(java.util.jar.Manifest.FastInputStream)) source' | wc
      0       0       0
```
I think the reason is the built-in class loader's Java code takes a different code path than the CDS shared class loading code ([systemDictionaryShared.cpp](https://github.com/openjdk/jdk/blob/5c0d985abf7ef89f9a035692fe9db3ef4001bc2c/src/hotspot/share/classfile/systemDictionaryShared.cpp#L687)). So we should exercise the CDS code during dump time to ensure that the classes used by CDS are always loaded.

File: src/hotspot/share/memory/metaspaceShared.cpp [hunk 1]
```diff
@@ -1356,6 +1356,8 @@ void MetaspaceShared::preload_and_dump(TRAPS) {
 
     HeapShared::init_for_dumping(THREAD);
 
+    // create a dummy manifest to cause more classes loaded
```
Comment (iklam) [inline src/hotspot/share/memory/metaspaceShared.cpp:1359]: How about: `// exercise the manifest processing code to ensue classes used by CDS are always archived`?

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 2]
```diff
@@ -671,30 +687,17 @@ Handle SystemDictionaryShared::get_shared_jar_manifest(int shared_path_index, TR
   Handle manifest ;
   if (shared_jar_manifest(shared_path_index) == NULL) {
     SharedClassPathEntry* ent = FileMapInfo::shared_path(shared_path_index);
-    long size = ent->manifest_size();
+    size_t size = (size_t)ent->manifest_size();
```
Comment (iklam) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:690]: size_t is unsigned, so I think you should change the following test to `if (size == 0)`

**Review by iklam — APPROVED** [review]

File: src/hotspot/share/memory/metaspaceShared.cpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/memory/metaspaceShared.cpp:1359]: s/ensue classes/ensure classes/

**Review by iklam — APPROVED** [review]

**Review by calvinccheung — APPROVED** [review]
Looks good.

**Pull request closed without merging**